### DISCUSSION TITLE
Objects of negative coordinates can be made in the lazer editor and they behave differently between stable and lazer

### DISCUSSION THREAD

**Discussion by cihe13375** [ppy/osu#31223] [Q&A]
Objects of negative coordinates can be made in the lazer editor by distance snap:
![1](https://github.com/user-attachments/assets/9c4f2fc8-b269-4af0-817e-14d2aa62419c)
![2](https://github.com/user-attachments/assets/c3b4b917-4327-401c-a5e9-522c9ed9ad1f)
these notes survives saving and actually works in lazer gameplay (the note 1 is at y=0, while note 4 is at y<0):
![3](https://github.com/user-attachments/assets/76a84aa6-86eb-47eb-bd8b-cca35baf1ef7)
they also survives a .osz export and can be imported to stable.
[unknown artist - unknown title (Guest).osz.zip](https://github.com/user-attachments/files/18219422/unknown.artist.-.unknown.title.Guest.osz.zip)
however, stable gameplay displays the object at y=0:
![4](https://github.com/user-attachments/assets/1226a47b-4e19-4ccf-89d4-28ae6be90e8a)

This discrepancy implies two issues:
1. we probably want to disallow negative coordinates in the editor (cannot be created and are automatically moved to x=0/y=0 at save), because they will behave differently when played in stable which is likely not desired
2. I don't know if any ranked map has such objects, if they have then the scores set on lazer and stable would be incomparable because the maps are essentially different (though they are already incomparable anyway due to sliderhead accuracy).

Edit:
x>512 has the same issue.
y>384, however, somehow works in stable (stable editor does not forcefully move them back and they are displayed "correctly" in stable gameplay), so maybe it does not need to be fixed.

**Comment by cihe13375** [comment]
As now lazer supports map submission, I want to bump the visibility of this, because it could actually happen that someone submits maps with x<0, y<0, or x>512 objects from lazer, and the map is ranked. Indeed I won't be surprised if some ranked maps already have this problem. (the distance snap method above is already fixed, but there are still other ways to create objects with such coordinates.)

As lazer is open-sourced, I don't think any fix in the editor itself could fundamentally solve it. It need to be fixed from the gameplay side, either by allowing them in stable gameplay, or clamping them in lazer gameplay.

**Reply by bdach** [reply]
> but there are still other ways to create objects with such coordinates

These should be reported.

> or clamping them in lazer gameplay

I can't find a citation for this but I'm like 99% sure this is not a solution because there are aspire maps that depend on objects with coordinates out of bounds and if we clamp then we get screeched at for "ruining aspire".

**Comment by peppy** [comment]
@Hiviexd can you give some attention to this, and make sure we have verify tab checks for this (and confirm that the BNs are using the verify tab or AiMod and would be made aware of this)?

I don't think we'll be taking immediate action for/against this, but we should make sure maps don't get ranked with such issues in the mean time.

**Reply by Hiviexd** [reply]
Adding a verify check sounds like the best approach here, will work on something.

IMO we shouldn't outright clamp out-of-bound objects due to aspire breakage as bdach said, tho fixing this specific edge case with distance snap at some point sounds reasonable.

**Reply by Hiviexd** [reply]
https://github.com/ppy/osu/pull/36916

**Comment by cihe13375** [comment]
> These should be reported.

I can recall two methods for now (plus the edit externally power):
1. create a slider whose tail is beyond 512x384, and Ctrl+G
2. place two circles at one border of the 512x384 area (let's say, (x,y) = (0,100) and (0,200)), select them and drag the rotation handle
